### PR TITLE
Patch fix several bugs

### PR DESCRIPTION
**Pull request by dstallenberg**
Fixes the following bugs
- The iterate function was not awaited while it is an async function
- The metric manager did not set up the correct metrics
- The cfg program and cfg functions did not allow for contracted cfg's
- The edge contraction algorithm should not contract entry/exit nodes
- The exit event was called on the main eventemitter causing the program to exit
- rmDirSync does not work when the directory is not empty
- The cfg getincomingedges/getoutgoingedges functions where exposing internal structure making the cfg object mutable
- The distance calculation function was incorrect

### CODE REVIEW

File: libraries/cfg-core/lib/graph/ControlFlowGraph.ts [hunk 1]
```diff
@@ -91,7 +103,7 @@ export class ControlFlowGraph<S> {
    */
   private getIncomingEdgesMap(): Map<string, Edge[]> {
     const map = new Map<string, Edge[]>();
-    for (const edge of this._edges) {
+    for (const edge of this.edges) {
```
Comment (mitchellolsthoorn) [inline libraries/cfg-core/lib/graph/ControlFlowGraph.ts:106]: Why do we use the public getter and not the internal structure?

File: libraries/cfg-core/test/algorithms/edgeContraction.test.ts [hunk 2]
```diff
@@ -62,19 +62,19 @@ describe("CFG edge contraction", function () {
     cfg = new ControlFlowGraph(nodeRoot, nodeExit, nodeExit, nodes, edges);
   });
 
-  it("Two edges should be contracted", () => {
+  it("one edge should be contracted", () => {
```
Comment (mitchellolsthoorn) [inline libraries/cfg-core/test/algorithms/edgeContraction.test.ts:65]:
```suggestion
  it("One edge should be contracted", () => {
```

File: libraries/core/lib/metaheuristics/SearchAlgorithm.ts [hunk 3]
```diff
@@ -148,7 +148,7 @@ export abstract class SearchAlgorithm<T extends Encoding> {
       );
 
       // Start next iteration of the search process
-      this._iterate(budgetManager, terminationManager);
+      await this._iterate(budgetManager, terminationManager);
```
Comment (mitchellolsthoorn) [inline libraries/core/lib/metaheuristics/SearchAlgorithm.ts:151]: Why is this necessary?

File: libraries/core/lib/objective/BranchObjectiveFunction.ts [hunk 4]
```diff
@@ -89,7 +116,8 @@ export class BranchObjectiveFunction<
     const branchDistance = this.branchDistance.calculate(
       closestCoveredBranchTrace.condition_ast,
       closestCoveredBranchTrace.condition,
-      closestCoveredBranchTrace.variables
+      closestCoveredBranchTrace.variables,
+      trueOrFalse
```
Comment (mitchellolsthoorn) [inline libraries/core/lib/objective/BranchObjectiveFunction.ts:120]: I assume this is to restore the opposite branch calculation without having to select the smallest value?

File: libraries/cfg-core/lib/graph/ControlFlowGraph.ts (hunk 1 above)
Comment (dstallenberg) [reply libraries/cfg-core/lib/graph/ControlFlowGraph.ts:106]: No clue, fixed it!

File: libraries/cfg-core/test/algorithms/edgeContraction.test.ts (hunk 2 above)
Comment (dstallenberg) [reply libraries/cfg-core/test/algorithms/edgeContraction.test.ts:65]: Sharp, fixed it.

File: libraries/core/lib/metaheuristics/SearchAlgorithm.ts (hunk 3 above)
Comment (dstallenberg) [reply libraries/core/lib/metaheuristics/SearchAlgorithm.ts:151]: Because iterate is an async function (in the implementation in EvolutionaryAlgorithm.ts) and otherwise we would not wait for the function to finish.

File: libraries/core/lib/objective/BranchObjectiveFunction.ts (hunk 4 above)
Comment (dstallenberg) [reply libraries/core/lib/objective/BranchObjectiveFunction.ts:120]: Well, I came to the conclusion that selecting the smallest value does not work because if you cover the true branch (i.e. distance = 0) then the false branch always gives a non-zero branch distance. So if you take the smallest value you always end up with 0.

This allows us to calculate for a specific branch.

File: libraries/core/lib/metaheuristics/SearchAlgorithm.ts (hunk 3 above)
Comment (mitchellolsthoorn) [reply libraries/core/lib/metaheuristics/SearchAlgorithm.ts:151]: Wouldn't this have caused problems in the past or has it only recently been modified to be async?

File: libraries/core/lib/objective/BranchObjectiveFunction.ts (hunk 4 above)
Comment (mitchellolsthoorn) [reply libraries/core/lib/objective/BranchObjectiveFunction.ts:121]: So, my previous assumption that this logic should be checked turned out to be true

**Review by mitchellolsthoorn — APPROVED** [review]

File: libraries/core/lib/metaheuristics/SearchAlgorithm.ts (hunk 3 above)
Comment (dstallenberg) [reply libraries/core/lib/metaheuristics/SearchAlgorithm.ts:151]: No, I think I accidentally removed it in one of the refactors, because I remember it used to await correctly. I just added it back now.
